src/zip.h

```diff
@@ -494,6 +495,9 @@ extern ZIP_EXPORT ssize_t zip_stream_copy(struct zip_t *zip, void **buf,
 /**
```

Comment:
We *should not* deallocate memory allocated by other library/thread. if we need to free pre-allocated memory we either should add allocator/deallocator functionality or let people pass deallocator which will be called by this function.
It can be problematic, e.g.  if people allocate memory in C++ by `new` and we deallocate memory with `free`

Reply:
Yeah, it felt kind of hacky to me too. Maybe what I do is just create a copy if stream is not NULL.

Reply:
I would not copy. Just push the responsibility for the user and take `free_cb` as function pointer, which will be called (if not NULL) to deallocate memory. Then we are pretty sure that user's functions allocated/deallocated memory (not zip library).

Reply:
The problem is that if I don't make a copy of the stream, then internally, reallocs may happen that result in the original stream pointer getting freed anyway. So even if I did push changes that made the API breaking change of updating zip_stream_open to take a free callback function, I don't think the double free problem would be completely solved. 

Given that the provided stream may mutate as a result of realloc, I think the safest and simplest change here is to copy the stream into another buffer.

Reply:
> The problem is that if I don't make a copy of the stream, then internally, reallocs may happen that result in the original stream pointer getting freed anyway. So even if I did push changes that made the API breaking change of updating zip_stream_open to take a free callback function, I don't think the double free problem would be completely solved.
> 
> Given that the provided stream may mutate as a result of realloc, I think the safest and simplest change here is to copy the stream into another buffer.

So far, I do not see the point to copy the stream (it can be huge, so it will consume 2x memory). Moreover we already have `zip_stream_copy` function (where user is obligated to free memory).

Do you have any workflow scenario (with realloc) when the _double free_  problem happens?
In terms of tests, I think we can use two `zip_t` pointers and `close`/`free` them at the end (in correct order), to avoid memory leaks.

Reply:
The realloc may happen any time someone passes a stream to zip_open_stream and then adds a file to the zip that was big enough that the realloc moved the memory that stream originally pointed to to another address.

I agree that copying the stream is not ideal if the stream may be gigabytes. I think part of fixing this issue would require passing the stream pointer by reference (e.g. `int zip_stream_open(uint8_t **ppstream, ...)`) so that the realloc memory shift won't matter matter as much. The user would then almost not have to call zip_stream_copy at that point to get the final zip, but there's some zip finalization that has to take place that isn't exposed to the API user (`zip_archive_finalize`). The other part of addressing this is to create a boolean that reflects whether the memory that the zip API is operating on was originally allocated by the user and to not free it if that was the case. 

It may take me a few to do the above, but let me know your thoughts.

Reply:
> The realloc may happen any time someone passes a stream to `zip_open_stream` and then adds a file to the zip that was big enough that the realloc moved the memory that stream originally pointed to to another address.
> 

Ok, I understand what you mean. That's why to kept workflows safe and separated `delete` mode from `write/append`.
Frankly speaking, I'm not super convinced that it's worth to redesign the API, or copy the stream, ...
Personally, I do not think it's a big deal to open the stream twice. First time for delete, the second time for write/append.